cogs/roll.py

```diff
@@ -0,0 +1,40 @@
+"""Command to imitate an ingame dice roll. Can roll randomly, or return the summary of the roll."""
+from hagadias.helpers import DiceBag
+import logging
+
+from discord import Colour, Embed
+from discord.ext.commands import Cog, Context, command
+
+log = logging.getLogger('bot.' + __name__)
+
+
+class Roll(Cog):
+    """Takes a dice string argument and either returns the stats for it or use as a base to roll.
+    Syntax: ?dice (dice string)"""
+
+    @command()
+    async def dice(self, ctx: Context, *args):
+        """Returns the (min, max) Average (average)."""
+        log.info(f'({ctx.message.channel}) <{ctx.message.author}> {ctx.message.content}')
+        val = ' '.join(args)
+        avg = DiceBag(val).average()
+        min = DiceBag(val).minimum()
+        max = DiceBag(val).maximum()
+        msg = f'{val}: ({min}, {max}), average {avg}'
+        embedded_msg = Embed(colour=Colour(0xf403f), description=msg)
+        return await ctx.send(embed=embedded_msg)
+
+    """Randomly rolls based on the string provided using Hagadias's dice roll helper.
+    Syntax: ?roll (dice string)
+    The dice string can only contain [0-9][d-+]. Unlike the Qud wiki module, it parses
+    ranges ex. "1-5" as subtraction."""
+    @command()
+    async def roll(self, ctx: Context, *args):
```

Comment:
`?roll 1d0` causes an exception and `?roll 1000000000d100` freezes up the bot. Rolling `nd1` seems to always produce n.

This might be a good place to have the `@command` function handle only the *Discord* input and output, and delegate the *dice* input and output to a helper function that can be rigorously tested by pytest (Discord commands are notoriously difficult to test.)

Reply:
There seem to be some issues with the dicebag in hagadias that we simply didn't encounter because the inputs were trusted before, but now we're moving it into a potentially untrusted environment (Discord).

It probably needs a mixture of sanitizing on the Discord end (denial of service) and sanity checking on the hagadias end (exceptions).

Reply:
We should also catch ValueErrors, since DiceBag can throw those and we can give a useful reply

Reply:
Since hagadias does this checking now, I added try/excepts and return the value error messages. Will be going back to hagadias to add another check because `-` is still a valid string.